src/Streamly/Internal/Data/Array/Foreign/Mut/Types.hs

```diff
@@ -523,6 +524,33 @@ read = Unfold step inject
             return $ D.Yield x
                 (ReadUState fp (p `plusPtr` sizeOf (undefined :: a)))
 
+-- | Unfold an array into a stream in reverse order.
+--
+-- /Internal/
+{-# INLINE_NORMAL readRev #-}
+readRev :: forall m a. (Monad m, Storable a) => Unfold m (Array a) a
+readRev = Unfold step inject
+    where
+
+    inject (Array fp end _) =
+        let p = end `plusPtr` negate (sizeOf (undefined :: a))
+         in return $ ReadUState fp p
+
+    {-# INLINE_LATE step #-}
+    step (ReadUState fp@(ForeignPtr start _) p) | p < Ptr start =
+        let x = unsafeInlineIO $ touchForeignPtr fp
+        in x `seq` return D.Stop
+    step (ReadUState fp p) = do
+            -- unsafeInlineIO allows us to run this in Identity monad for pure
+            -- toList/foldr case which makes them much faster due to not
+            -- accumulating the list and fusing better with the pure consumers.
+            --
+            -- This should be safe as the array contents are guaranteed to be
+            -- evaluated/written to before we peek at them.
+            let !x = unsafeInlineIO $ peek p
+            return $ D.Yield x
+                (ReadUState fp (p `plusPtr` negate (sizeOf (undefined :: a))))
```

Comment:
```suggestion
            let next = ReadUState fp (p `minusPtr` sizeOf (undefined :: a))
            return $ D.Yield x next
```

Reply:
`minusPtr` is not what you think. `minusPtr :: Ptr a -> Ptr b -> Int `